CIP-0120? | Constitution specification (#796)
cardano-foundation/CIPs: cip-constitution-tech/README.md

---
CIP: ?
Title: Constitution form
Category: Metadata
Status: Proposed
Authors:
    - Ryan Williams <ryan.williams@intersectmbo.org>
Implementors: []
Discussions:
    - https://github.com/cardano-foundation/CIPs/pull/796
Created: 2024-03-19
License: CC-BY-4.0
---

## Abstract

> Cardano's minimum viable governance model as described within [CIP-1694 | A First Step Towards On-Chain Decentralized Governance](https://github.com/cardano-foundation/CIPs/blob/master/CIP-1694/README.md) introduces the concept of a Cardano constitution.

Reviewer: This is fine here because "constitution" isn't any specific document i.e. a common noun.

Author: understood, thank you!

Although CIP-1694 gives no definition to the constitution's content or form.

> This proposal aims to describe a standardized technical form for the Cardano constitution.
> Aiming to enhance the accessibility and safety of the document.

Reviewer: ```suggestion
This proposal aims to describe a standardized technical form for the Cardano Constitution to enhance the accessibility and safety of the document.
```
With the term "the" it defaults to THE Constitution, but could be changed to "a constitution" and demoted to a common noun if intended to refer to ALL documents (which might make more sense here).

> **Note:** This proposal only covers the technical form of the constitution, this standard is agnostic to the content of the constitution.

## Motivation: why is this CIP necessary?

> CIP-1694 defines the on-chain anchor mechanism used to link the off-chain constitution document to on-chain actions.

Reviewer: ```suggestion
CIP-1694 defines the on-chain anchor mechanism used to link the off-chain Constitution document to on-chain actions.
```
If this is the "live" constitution, we have to be talking about the one & only Constitution in effect at the time (?).

> This mechanism was chosen due to its simplicity and cost effectiveness, moving the potentially large Cardano constitution off-chain, leaving only a a hash digest and URI on-chain.

Reviewer: ```suggestion
This mechanism was chosen due to its simplicity and cost effectiveness, moving the potentially large Cardano constitution off-chain, leaving only a hash digest and URI on-chain.
```

Author: good catch, thanks

> This is the extent to which CIP-1694 outlines the Cardano constitution, CIP-1694 does not provide suggestions around hashing algorithm, off-chain storage best practices or use of rich text styling.

Reviewer: ```suggestion
This is the extent to which CIP-1694 outlines the Cardano Constitution: CIP-1694 does not provide suggestions around hashing algorithm, off-chain storage best practices or use of rich text styling.
```
I will stop marking what I think is this same capitalisation error until we have some consensus about the suggested guidelines (which could also become rules in the CIP) & just mark incidental changes below.

> By formalizing the form of the constitution and it's iterations, we aim to promote it's longevity and accessibility.
> This is essential to ensure the effectiveness of the CIP-1694 governance model.

Reviewer: ```suggestion
By formalizing the form of the constitution and its iterations, we aim to promote its longevity and accessibility.
This is essential to ensure the effectiveness of the CIP-1694 governance model.
```

> This standard will impact how ada holders read the constitution but the main stakeholders for this are the tool makers who wish to read, render and write the constitution.

Reviewer: ```suggestion
This standard will impact how Ada holders read the Constitution but the main stakeholders for this are the tool makers who wish to read, render and write a constitution.
```
or "a constitution" throughout, if people are being encouraged to think in this way (?).  Having these two spellings in this one sentence might show by example the conceptual difference between "THE Constitution" and a set of potential constitutions.

### Safety

Without describing best practices for the form and handling of the constitution, we risk the constitution document being stored in an insecure manner.

> By storing the constitution on a decentralized platform, we can ensure it's immutability and permissionless access.

Reviewer: ```suggestion
By storing the constitution on a decentralized platform, we can ensure its immutability and permissionless access.
```

This is a step to improve the longevity and accessibility of each constitution iteration.

### Interoperability

By defining a file extension and formatting rules for the constitution we ensure that tooling reading and writing the constitution will be interoperable.
Furthermore we aim to make the role of constitution iteration comparison tools easier, by minimizing formatting and style changes between iterations.
This will reduce compatibility issues between tools, promoting the accessibility of the constitution.

### Usability

Rich text formatting greatly enhances the readability of text, especially in large complex documents.
Without the ability to format text, it could easily become cumbersome to read, negatively effecting the accessability of the Cardano constitution.

## Specification

> The following specification SHOULD be applied to all constitution iterations.

Reviewer: Excellent presentation of the common noun "constitution" as part of a multiplicity extended in _time_ as opposed to space.

Author: Understood, thanks Robert

This standard could be augmented in the future via a separate CIP which aims to replace this one.

### Sentences

> The constitution text MUST only contain at maximum one sentence per line.

Reviewer: ```suggestion
The constitution text MUST only contain a maximum of one sentence per group of lines delimited by one blank line between these groups.
```
I have to conclude this based this current second "example" being valid (not "one sentence per line" but rather one sentence in the group of lines that's combined into a single flowing line by Markdown):
```
This is a long sentence and I have valid reasons for it being so long,
such as being an example of a long sentence.
```

Author: good catch, thank you!

Reviewer: see re-opening of this discussion: https://github.com/cardano-foundation/CIPs/pull/796#pullrequestreview-2085822746

such as being an example of a long sentence.
```

It is recommended that each line is ended by either a full stop or comma.

### File Type

The constitution file MUST be a text file named `constitution.txt`.

### Hashing

When supplying a constitution hash digest to chain, the algorithm used MUST be Blake2b-256.
Before creating a hash digest the constitution plain text MUST be in its raw text, including any [Rich Text Formatting](#rich-text-formatting) related characters.

### Storage

The each ratified constitution MUST be stored, immutably on a distributed storage mechanism such as IPFS.
Where backups can be easily made in a permissionless manner by interested parties.
This storage platform MUST be easily accessible, with strong tooling support.
Authors SHOULD NOT specify the use of centralized gateways to access the constitution plain text.

// only ascii

### Rich Text Formatting

The constitution text MAY include a strict subset of rich text styling as defined in this specification.
Tooling rendering the constitution SHOULD recognize these and render them faithfully.

#### Headers

Headers are denoted via a hashtag character `#` followed by a space ` `.

> There are six levels to headers with, each being defined via a extra `#`.

Reviewer: ```suggestion
There are six levels of headers, with each lower level set via an added `#`.
```

Reviewer: There is probably a standard way of describing this which can be cobbled from Markdown and/or W3C documentation.  I proposed the above since I thought it made the statement more accurate with a minimum amount of change.

Reviewer: Just checked the [Markdown Guide](https://www.markdownguide.org/basic-syntax/#headings) and [GitHub Docs](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#headings) and neither contains a description that's both concise and complete, so the above is the best I can offer. 😖

Headers are ended via a line break.
Headers SHOULD be followed below by a blank line.
Headers SHOULD not be preceded by whitespace.

The lower the number of `#` the larger order the text SHOULD be rendered.

Example:
```md
# H1

## H2

### H3

#### H4

##### H5

###### H6

```

If text is in a header no other formatting can be applied.

#### Emphasis

Emphasis is applied to text between single or double asterisks, without space between asterisks and text.
Italicized emphasis is shown via single asterisk (`*`).
Bold emphasis is shown via double asterisks (`**`).

Emphasis cannot span multiple lines.

Examples:
```md
Emphasis, aka italics, with single *asterisks*.

Strong emphasis, aka bold, with double **asterisks**.
```

The text contained within headings can be emphasized.

Both italicized and bold cannot be applied to the same text.

#### Code and Syntax Highlighting

Texted can be highlighted as code, when encased without spaces by backticks.

> This MUST not contain line breaks.

Reviewer: @Ryun1 can we be sure a constitution will not contain code blocks?

Author: not if it follows this CIP 😆

Reviewer: I guess I should have said, "How could we be sure there will _not_ be a convincingly proposed constitution that contains code blocks?" 🤔

Author: I dont see it being necessary
but also cant predict it

Reviewer: maybe... but I wasn't thinking in terms of missing markdown features... rather, possible instances of a constitution in which a code block could indicate some vital feature of governance that inline code could not.

Inline code tags would be generally used for machine readable terms... especially blockchain parameters... but what if a simple procedure needed to be indicated?  Not something technical like Plutus or Javascript, but pseudocode as for a simple flowchart or high level algorithm?  Would all constitutions be required to indicate these procedures with nested lists and such?

Example:
```md
Inline `code` has `back-ticks around` it.
```

The text contained within headings or emphasis cannot be highlighted as code.

#### Unordered Lists

// todo

#### Line Breaks / Paragraphs

To create paragraphs, use a blank line to separate one or more lines of text.

Examples:
```md
Here's a line for us to start with.

This line is separated from the one above by two newlines, so it will be a *separate paragraph*.

This line is also a separate paragraph, but...
This line is only separated by a single newline, so it's a separate line in the *same paragraph*.
```

### Best Practices
- when hash doesnt match
- when unrecognised formatting/style
- Keep lines left-aligned without unneeded whitespace or tabs

> - tabs vs spaces

Reviewer: Not editing this section because it looks like a draft (@Ryun1 please let us know if otherwise)

Author: Yup, still a draft
just going to write out some general good practices/advice

> ## Rationale: how does this CIP achieve its goals?

Reviewer: ```suggestion
## Rationale: how does this CIP achieve its goals?
```
as before, this section looks "drafty" so not nitpicking about it (@Ryun1 please let us know when this is in full form & ready to review).

Author: Yup, still in draft
I want to sure up the spec before writing the Rationale

Just like CIP-1694, we aim to define a minimal viable set of standards to ensure governance can be achieved.

### Sentences
- improve the readability of diff views

### Versioning
- why not build in options in hashalgo (like CIP100)
- why a whole replacement?

### File Type
- just simple and easy

### Hashing
- we choose blake2b-256 as it is a standard for hashing in Cardano
- standard, lots of tools already use it 

### Storage
- The constitution is the most important document for Cardano (?), ensuring its permissionless accessibly is paramount
- Storing each iteration immutably is just as important, to allow access of past constitutions
- IPFS is well known already with good tooling support -> improve accessibility

### Rich Text Formatting
- why not all markdown?
- wanted to keep a bare minimum spec, no images or videos
- also dont want people to be able to attack consumers by attacks utilizing the HTML
- wanted to allow standard markdown libs to work

> ## Open Questions

Reviewer: Since this is only a general section for the CPS format I also assume these are author's notes and/or an editorial checklist so please @Ryun1 indicate if this information is to be kept in the final version of the document.

Author: Yeah, more of an editorial check list
I may or may not keep it as an optional section in the final version

- [x] How can we support multi-languages?
  - The Cardano constitution will be in English, but we will add best practice guidelines via [Best Practices](#best-practices).
- [ ] Should we specify any standardization for the proposal policy?
- [x] How can we add page breaks?
  - We wont, instead we will prioritize a minimum set of rich text formatting. We can provide some guidance via [Best Practices](#best-practices).
- [x] Do we want a mechanism for specifying authors? (similar to CIP-100)
  - No, as CIP-100 compliant metadata can be supplied at time of constitution update.
- [ ] What should we name the constitution file? we could embed some nice naming or metadata.

## Path to Active

### Acceptance Criteria

> - [ ] This standard is followed for the interim Cardano constitution

Reviewer: ```suggestion
- [ ] This standard is followed for the interim Cardano Constitution
```
Last nitpick this time around I promise: there will be only one interim constitution so it should be capitalised as proper noun.  It is ambiguous (in my understanding) whether Interim should also be capitalised... perhaps having political implications if it were so & therefore better to leave in lowercase. 😜

Author: ALL NITPICKS APPRECIATED 😎

Reviewer: thanks... @Ryun1 do you know if the working groups would prefer "interim" by itself as a relatively insignificant modifier, or including it capitalised in the title as "Interim Cardano Constitution"?

Author: I dont think there is official language defined, happy to use what you suggest

Reviewer: There will be only one Interim Constitution, so it would be grammatically correct to capitalise it. 🤓

---
CIP: ?

> Title: Constitution form

Author: @rphair you are very good at naming, please let me know what you think of this title vs "Constitution technical form"

Reviewer: My first choice would have been `Constitution Process` but I think too many people would assume this meant either an _approval_ process or a Constitution mandated governance process.

`form` is a good choice but it doesn't include the mechanics like hashing or off-chain storage.  The general engineering term `design` would refer to both form _and_ function so my current best idea would be `Constitution Design` (with the second term also capitalised, since there will only be one of these 😜).

Reviewer: My first choice would have been `Constitution Process` but I think too many people would assume this meant either an *approval* process or a Constitution mandated governance process.

`form` is a good choice but it doesn't include the mechanics like hashing or off-chain storage. The general engineering term `design` would refer to both form *and* function so my current best idea would be `Constitution Design` (with the second term also capitalised, since there will only be one of these 😜).

Author: `Constitution Design` I feel implies more of the structure of the content of the Constitution

Author: Settled on `Constitution specification`, thanks for the suggestion @microproofs

By storing the constitution on a decentralized platform, we can ensure its immutability and permissionless access.
This is a step to improve the longevity and accessibility of each constitution iteration.

### Interoperability

By defining a file extension and formatting rules for the constitution we ensure that tooling reading and writing the constitution will be interoperable.
Furthermore we aim to make the role of constitution iteration comparison tools easier, by minimizing formatting and style changes between iterations.
This will reduce compatibility issues between tools, promoting the accessibility of the constitution.

### Usability

Rich text formatting greatly enhances the readability of text, especially in large complex documents.
Without the ability to format text, it could easily become cumbersome to read, negatively effecting the accessability of the Cardano constitution.

## Specification

The following specification SHOULD be applied to all constitution iterations.
This standard could be augmented in the future via a separate CIP which aims to replace this one.

### Terminology

> 

Author: > If you want I will prepare that as a code diff.

@rphair
I do think these rules are true and helpful
Please add these rules into this section 💪

Reviewer: @Ryun1 added under `### Terminology`:

---

#### Capitalisation of key terms

To avoid unnecessary edits, and therefore checksum changes, constitution authors MUST follow the following standard English capitalisation rules unless a translation language indicates otherwise:

`Constitution` vs. `constitution`

- The Constitution in effect — either the "initial" one or any updated version — is unique and therefore capitalised ("Constitution") as a proper noun.
- Draft or proposed constitutions are not unique & therefore are not capitalised ("constitution") as a common noun.
- "Cardano Constitution" is a very specific proper noun phrase (also a title) and so each word is capitalised.
- The phrase "the Constitution", unless used non-specifically (e.g. "the constitution that voters prefer"), would generally be assumed to be _the_ Constitution and capitalised as a proper noun.

`Ada` vs. `ada` vs. `ADA`

- `Ada` is the currency, while `ada` indicates units of that currency (e.g. "Ada holders can accumulate more ada to increase their influence.")
- `ADA` is the trading symbol (e.g. "Fluctuations in ADA might influence decisions about the Treasury.")

Author: honestly a Cardano technical terminology glossary, would be a useful thing
has this been done before?

Reviewer: @Ryun1 the closest thing I know of that type is this page on the Dev Portal (maybe @katomm knows of another?): https://developers.cardano.org/docs/portal-style-guide/#editorial-style-guide

- when unrecognised formatting/style
- Keep lines left-aligned without unneeded whitespace or tabs
- tabs vs spaces

## Rationale: how does this CIP achieve its goals?

Just like CIP-1694, we aim to define a minimal viable set of standards to ensure governance can be achieved.

### Sentences
- improve the readability of diff views

### Versioning
- why not build in options in hashalgo (like CIP100)
- why a whole replacement?

### File Type
- just simple and easy

### Hashing
- we choose blake2b-256 as it is a standard for hashing in Cardano
- standard, lots of tools already use it 

### Storage
- The constitution is the most important document for Cardano (?), ensuring its permissionless accessibly is paramount
- Storing each iteration immutably is just as important, to allow access of past constitutions
- IPFS is well known already with good tooling support -> improve accessibility

### Rich Text Formatting
- why not all markdown?
- wanted to keep a bare minimum spec, no images or videos
- also dont want people to be able to attack consumers by attacks utilizing the HTML
- wanted to allow standard markdown libs to work

## Open Questions
- [x] How can we support multi-languages?

>   - The Cardano constitution will be in English, but we will add best practice guidelines via [Best Practices](#best-practices).

Reviewer: Since the `<link>` attribute from HTML metadata is ignored by Markdown:
- https://gist.github.com/rphair/144bab3002a277fa404235e2f1c4d798
- https://gist.github.com/rphair/1352af824be2fdc5b5e562df465af0e5

... we could include, as the first line of the Constitution document:
```
<link rel="alternate" hreflang="fr" href="#" />
```

... to specify the language, and possibly also:
```
<link rel="alternate" hreflang="x-default" href="#" />
```

... to indicate that this document is in the default language: so that constitution authors could specify that a submitted document is in _its_ default language.  Otherwise a default language of `en` could be specified in the CIP.

It might also be helpful to set the `href=` attribute to the IPFS hash instead of `#`... but that would depend upon the overall Constitution process.  [W3C says](https://www.w3.org/TR/2011/WD-html5-author-20110809/the-link-element.html) it must be set to a valid URI; the `rel=` attribute is also required and is [generally](https://ahrefs.com/blog/hreflang-tags/) set to `alternate` for `<link>`s in HTML headers.

cc @Hornan7 @Crypto2099 @Quantumplation

Author: I do like the idea, and had considered something like this for indicating the hashing algorithm.

I think id prefer not to do this, just for simplicity of tooling.

---

> CIP: ?

Reviewer: ```suggestion
CIP: 120
```

> Title: Constitution form

Author: ```suggestion
Title: Constitution specification
```

> Title: Constitution form

Author: ```suggestion
Title: Constitution specification
```
